File: go/pkg/hey/collections.go
package hey

import (
	"context"
	"fmt"
	"net/url"
	"strconv"

	"github.com/basecamp/hey-sdk/go/pkg/generated"
)

// CollectionsService handles collections — shared threads gathered under one name.
type CollectionsService struct {
	client *Client
}

// NewCollectionsService creates a new CollectionsService.
func NewCollectionsService(client *Client) *CollectionsService {
	return &CollectionsService{client: client}
}

// CreateCollectionParams contains the parameters for creating a collection.
type CreateCollectionParams struct {
	// Name is what the collection is called.
	Name string
	// Summary is the optional blurb shown under the name.
	Summary string
	// AccountID picks which account owns it. Zero leaves the server to pick your first.
	AccountID int64
}

// UpdateCollectionParams contains the parameters for editing a collection.
// Empty fields are left alone.
type UpdateCollectionParams struct {
	Name    string
	Summary string
}

// CollectionPage contains one page of a collection and its pagination state.
type CollectionPage struct {
	Collection *generated.CollectionWithPostings
	NextPage   string
	TotalCount int
}

// List returns your collections.
func (s *CollectionsService) List(ctx context.Context) (result *generated.ListCollectionsResponseContent, err error) {
	op := OperationInfo{
		Service: "Collections", Operation: "ListCollections",
		ResourceType: "collection", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().ListCollectionsWithResponse(ctx)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = resp.JSON200
		return nil
	})
	return result, err
}

// Get returns a collection and the active, accessible threads in its requested page.
func (s *CollectionsService) Get(ctx context.Context, collectionID int64, params *generated.GetCollectionParams) (*generated.CollectionWithPostings, error) {
	page, err := s.GetPage(ctx, collectionID, params)
	if err != nil || page == nil {
		return nil, err
	}
	return page.Collection, nil
}

// GetPage returns a collection page with its next cursor and total posting count.
func (s *CollectionsService) GetPage(ctx context.Context, collectionID int64, params *generated.GetCollectionParams) (result *CollectionPage, err error) {
	op := OperationInfo{
		Service: "Collections", Operation: "GetCollection",
		ResourceType: "collection", IsMutation: false, ResourceID: collectionID,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().GetCollectionWithResponse(ctx, collectionID, params)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = &CollectionPage{Collection: resp.JSON200}
		if resp.HTTPResponse != nil {
			result.TotalCount, _ = strconv.Atoi(resp.HTTPResponse.Header.Get("X-Total-Count"))
			result.NextPage = gearedPageFromLink(resp.HTTPResponse.Header.Get("Link"))
		}
		return nil
	})
	return result, err
}

// Update renames a collection or changes its summary.
func (s *CollectionsService) Update(ctx context.Context, collectionID int64, params UpdateCollectionParams) error {
	op := OperationInfo{
		Service: "Collections", Operation: "UpdateCollection",
		ResourceType: "collection", IsMutation: true, ResourceID: collectionID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		// Empty strings are omitted by the payload's omitempty tags, so unset
		// fields are left alone server-side.
		resp, err := s.client.genClient().UpdateCollectionWithResponse(ctx, collectionID, generated.UpdateCollectionRequestContent{
			Collection: generated.CollectionPayload{Name: params.Name, Summary: params.Summary},
		})
		if err != nil {
			return err
		}
		return CheckResponse(resp.HTTPResponse)
	})
}

// Create makes a new collection.
//
// HEY has no JSON endpoint for this — the form post answers with a redirect to the
// collections index rather than the new collection, so the id is not returned. List
// afterwards to find it.
func (s *CollectionsService) Create(ctx context.Context, params CreateCollectionParams) error {
	op := OperationInfo{
		Service: "Collections", Operation: "CreateCollection",
		ResourceType: "collection", IsMutation: true,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		values := url.Values{}
		values.Set("collection[name]", params.Name)
		if params.Summary != "" {
			values.Set("collection[summary]", params.Summary)
		}
		if params.AccountID != 0 {
			values.Set("account_id", fmt.Sprintf("%d", params.AccountID))
		}

		_, err := s.client.PostForm(ctx, "/collections", values)
		return err
	})
}

// AddTopic files a topic into a collection.
//
// HEY has no JSON endpoint for this — the form post answers with a redirect to the topic.
func (s *CollectionsService) AddTopic(ctx context.Context, topicID int64, collectionID int64) error {
	op := OperationInfo{
		Service: "Collections", Operation: "CreateTopicCollecting",
		ResourceType: "collecting", IsMutation: true, ResourceID: topicID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		_, err := s.client.PostForm(ctx, fmt.Sprintf("/topics/%d/collecting?collection_id=%d", topicID, collectionID), url.Values{})
		return err
	})
}

// RemoveTopic takes a topic back out of a collection.
//
// HEY has no JSON endpoint for this — the form post answers with a redirect to the topic.
// Shadowed topics are silently left alone.
func (s *CollectionsService) RemoveTopic(ctx context.Context, topicID int64, collectionID int64) error {
	op := OperationInfo{
		Service: "Collections", Operation: "DeleteTopicCollecting",
		ResourceType: "collecting", IsMutation: true, ResourceID: topicID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		_, err := s.client.DeleteForm(ctx, fmt.Sprintf("/topics/%d/collecting?collection_id=%d", topicID, collectionID))
		return err
	})
}

File: rust/hey-sdk/src/services/collections.rs
//! Making collections and filing threads into them, on top of the generated collection
//! routes.
//!
//! HEY serves no JSON endpoint for any of these, so each one is a browser form post.

use reqwest::Method;

use crate::error::Error;
use crate::generated::types::{CollectionPayload, UpdateCollectionRequestContent};
use crate::services::write_info;

pub use crate::generated::services::collections::*;

/// What a new collection is made of.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct CreateCollectionParams {
    pub name: String,
    /// The blurb shown under the name.
    pub summary: Option<String>,
    /// The account that owns it. `None` leaves HEY to pick your first.
    pub account_id: Option<i64>,
}

/// What an edit changes about a collection. A field left unset — `None` or empty — is left
/// off the wire, and HEY leaves what a request does not name alone.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpdateCollectionParams {
    pub name: Option<String>,
    pub summary: Option<String>,
}

impl<'a> Collections<'a> {
    /// Makes a collection.
    ///
    /// The form post answers with a redirect to the collections index rather than to the
    /// collection it made, so the new collection's id does not come back.
    /// [`Collections::list`] afterwards is how to find it.
    pub async fn create(&self, params: &CreateCollectionParams) -> Result<(), Error> {
        let account = params.account_id.map(|account_id| account_id.to_string());
        let mut fields = vec![("collection[name]", params.name.as_str())];
        if let Some(summary) = params.summary.as_deref().filter(|it| !it.is_empty()) {
            fields.push(("collection[summary]", summary));
        }
        if let Some(account) = &account {
            fields.push(("account_id", account.as_str()));
        }

        let mut operation = self.client().form(Method::POST, "/collections")?;
        operation.info(write_info(
            "Collections",
            "CreateCollection",
            "collection",
            None,
        ));
        operation.form(&fields);
        self.client().send_unit(operation).await
    }

    /// Renames a collection or changes its summary. The generated [`Collections::update`]
    /// takes the same request as a body.
    pub async fn update_collection(
        &self,
        collection_id: i64,
        params: &UpdateCollectionParams,
    ) -> Result<(), Error> {
        let body = UpdateCollectionRequestContent {
            collection: CollectionPayload {
                name: present(&params.name),
                summary: present(&params.summary),
            },
        };
        self.update(collection_id, &body).await
    }

    /// Files a topic into a collection.
    pub async fn add_topic(&self, topic_id: i64, collection_id: i64) -> Result<(), Error> {
        let mut operation = self
            .client()
            .form(Method::POST, &format!("/topics/{topic_id}/collecting"))?;
        operation.info(write_info(
            "Collections",
            "CreateTopicCollecting",
            "collecting",
            Some(topic_id),
        ));
        operation.query("collection_id", collection_id).form(&[]);
        self.client().send_unit(operation).await
    }

    /// Takes a topic back out of a collection. A shadowed topic is silently left alone.
    pub async fn remove_topic(&self, topic_id: i64, collection_id: i64) -> Result<(), Error> {
        let mut operation = self
            .client()
            .form(Method::DELETE, &format!("/topics/{topic_id}/collecting"))?;
        operation.info(write_info(
            "Collections",
            "DeleteTopicCollecting",
            "collecting",
            Some(topic_id),
        ));
        operation.query("collection_id", collection_id);
        self.client().send_unit(operation).await
    }
}

/// An empty string is no value, and is left off the wire like a `None` one — the omission
/// is what tells HEY to leave the field as it is.
fn present(value: &Option<String>) -> Option<String> {
    value.clone().filter(|value| !value.is_empty())
}
